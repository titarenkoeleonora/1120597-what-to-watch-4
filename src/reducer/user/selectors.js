import NameSpace from "../name-space.js";

export const getAuthorizationStatus = (state) => state[NameSpace.USER].authorizationStatus;

export const getAuthorizationInfo = (state) => state[NameSpace.USER].authorizationInfo;

export const getIsAuthorizing = (state) => state[NameSpace.USER].isAuthorizing;
