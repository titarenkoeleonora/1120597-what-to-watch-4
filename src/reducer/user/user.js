import {UserActionCreator} from "../actions/user-action-creator";
import {UserActionType} from "../actions/user-action-type";
import {extend} from "../../utils";
import createUserInfo from "../../adapter/user";

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  isAuthorizing: false,
  authorizationInfo: {
    id: 0,
    email: ``,
    name: ``,
    avatar: ``,
  }};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then((response) => {
        dispatch(UserActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(UserActionCreator.loadAuthorizationInfo(createUserInfo(response.data)));
      })
      .catch((err) => {
        throw err;
      });
  },

  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.login,
      password: authData.password,
    })
      .then((response) => {
        dispatch(UserActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(UserActionCreator.loadAuthorizationInfo(createUserInfo(response.data)));
        dispatch(UserActionCreator.isNotAuthorizing());
      })
      .catch(() => {
        dispatch(UserActionCreator.requireAuthorization(AuthorizationStatus.NO_AUTH));
      });
  },
};


const reducer = (state = initialState, action) => {
  switch (action.type) {
    case UserActionType.REQUIRED_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload,
      });
    case UserActionType.AUTHORIZATION_INFO:
      return extend(state, {
        authorizationInfo: action.payload
      });
    case UserActionType.IS_AUTHORIZING:
      return extend(state, {
        isAuthorizing: true,
      });
    case UserActionType.IS_NOT_AUTHORIZING:
      return extend(state, {
        isAuthorizing: false,
      });
  }

  return state;
};

export {reducer, initialState, AuthorizationStatus, Operation};
