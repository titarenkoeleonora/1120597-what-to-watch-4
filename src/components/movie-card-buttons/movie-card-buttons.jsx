import React from "react";
import PropTypes from 'prop-types';
import {Link} from "react-router-dom";
import {AppRoute} from "../../constants";
import history from "../../history";
import {AuthorizationStatus} from "../../reducer/user/user";
import {getAuthorizationStatus} from "../../reducer/user/selectors";
import {connect} from "react-redux";
import {Operation as DataOperation} from "../../reducer/data/data";

const MovieCardButtons = ({activeMovie, authorizationStatus, changeMovieIsFavorite, onPlayButtonClick}) => {
  const handleMovieListButtonClick = () => {

    return authorizationStatus === AuthorizationStatus.AUTH
      ? changeMovieIsFavorite(activeMovie)
      : history.push(AppRoute.LOGIN);
  };

  const addToMyList = (
    <button
      className="btn btn--list movie-card__button"
      type="button"
      onClick={() => handleMovieListButtonClick(true)}
    >
      <svg viewBox="0 0 19 20" width="19" height="20">
        <use xlinkHref="#add"></use>
      </svg>
      <span>My list</span>
    </button>
  );

  const removeFromMyList = (
    <button
      className="btn btn--list movie-card__button"
      type="button"
      onClick={() => handleMovieListButtonClick(false)}
    >
      <svg viewBox="0 0 18 14" width="18" height="14">
        <use xlinkHref="#in-list"></use>
      </svg>
      <span>My list</span>
    </button>
  );

  return (
    <div className="movie-card__buttons">
      <Link
        className="btn btn--play movie-card__button"
        onClick={() => {
          onPlayButtonClick();
          history.push(`${AppRoute.PLAYER}/${activeMovie && activeMovie.id}`);
        }}
        to={`${AppRoute.PLAYER}/${activeMovie && activeMovie.id}`}
      >
        <svg viewBox="0 0 19 19" width="19" height="19">
          <use xlinkHref="#play-s" />
        </svg>
        <span>Play</span>
      </Link>
      {activeMovie.isFavorite ? removeFromMyList : addToMyList}
      <a href="add-review.html" className="btn movie-card__button">Add review</a>
    </div>
  );
};

MovieCardButtons.propTypes = {
  activeMovie: PropTypes.object,
  onPlayButtonClick: PropTypes.func,
  authorizationStatus: PropTypes.string.isRequired,
  changeMovieIsFavorite: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  changeMovieIsFavorite(movie) {
    dispatch(DataOperation.changeFavoriteStatus(movie));
  }
});

export {MovieCardButtons};
export default connect(mapStateToProps, mapDispatchToProps)(MovieCardButtons);
