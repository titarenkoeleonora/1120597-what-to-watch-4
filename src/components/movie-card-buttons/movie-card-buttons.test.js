import React from "react";
import renderer from "react-test-renderer";
import {MovieCardButtons} from "./movie-card-buttons";
import {Router} from "react-router-dom";
import history from "../../history";
import {AuthorizationStatus} from "../../reducer/user/user";

const testMovie = {
  id: 0,
  title: `Movie title`,
  poster: ``,
  previewImage: ``,
  bgImage: ``,
  backgroundColor: ``,
  src: ``,
  previewVideoLink: ``,
  description: ``,
  rating: 8.2,
  votes: 240,
  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`, `Tony Revoloru`, `Tilda Swinton`, `Tom Wilkinson`, `Owen Wilkinson`, `Adrien Brody`, `Ralph Fiennes`, `Jeff Goldblum`],
  runTime: 99,
  genre: `Genre`,
  date: 2020,
  isFavorite: true,
};

it(`Correctly render`, () => {
  const tree = renderer
    .create(
        <Router history={history}>
          <MovieCardButtons
            activeMovie={testMovie}
            onPlayButtonClick={()=>{}}
            onChangeMovieIsFavorite={()=>{}}
            authorizationStatus={AuthorizationStatus.AUTH}
            onAddReviewClick={()=>{}}
          />
        </Router>
    );

  expect(tree).toMatchSnapshot();
});
