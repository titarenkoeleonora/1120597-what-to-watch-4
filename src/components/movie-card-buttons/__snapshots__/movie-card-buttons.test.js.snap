// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly render 1`] = `
<div
  className="movie-card__buttons"
>
  <a
    className="btn btn--play movie-card__button"
    href="/player/undefined"
    onClick={[Function]}
  >
    <svg
      height="19"
      viewBox="0 0 19 19"
      width="19"
    >
      <use
        xlinkHref="#play-s"
      />
    </svg>
    <span>
      Play
    </span>
  </a>
  <button
    className="btn btn--list movie-card__button"
    onClick={[Function]}
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 19 20"
      width="19"
    >
      <use
        xlinkHref="#add"
      />
    </svg>
    <span>
      My list
    </span>
  </button>
  <a
    className="btn movie-card__button"
    href="/films/:undefined/review"
    onClick={[Function]}
  >
    Add review
  </a>
</div>
`;
